plugins/search-backend-module-pg/src/PgSearchEngine/PgSearchEngineIndexer.ts

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { BatchSearchEngineIndexer } from '@backstage/plugin-search-backend-node';
+import { IndexableDocument } from '@backstage/search-common';
+import { DatabaseStore } from '../database';
+
+type IndexerOptions = {
+  batchSize: number;
+  type: string;
+  databaseStore: DatabaseStore;
+};
+
+export class PgSearchEngineIndexer extends BatchSearchEngineIndexer {
+  private store: DatabaseStore;
+  private type: string;
+
+  constructor(options: IndexerOptions) {
+    super({ batchSize: options.batchSize });
+    this.store = options.databaseStore;
+    this.type = options.type;
+  }
+
+  async index(documents: IndexableDocument[]): Promise<void> {
+    await this.store.transaction(async tx => {
+      await this.store.prepareInsert(tx);
+      await this.store.insertDocuments(tx, this.type, documents);
+      await this.store.completeInsert(tx, this.type);
+    });
+  }
```

Comment:
Is `documents` a batch? Because it that case this code (for postgres) won't work. `prepareInsert` has to be called before the first batch and `completeInsert` after the last batch, and all of them in the same transaction 🤔

Reply:
Ah-hah, makes sense.  I guess we'd need to create/prepare (and keep a reference to) the transaction on instantiation.  Just call `insertDocuments` in the `index()` method, and then call `completeInsert()` in a custom `_final()` implementation.

Reply:
Possibly there are tweaks to `DatabaseDocumentStore` that could simplify this, but I've got this working in a second iteration on the indexer.  Have a look!